diff -ru -x .gitignore ../abhinavsingh-JAXL-5829c3b/core/jaxl.parser.php ../../lib/php/jaxl/core/jaxl.parser.php
--- ../abhinavsingh-JAXL-5829c3b/core/jaxl.parser.php	2011-05-01 07:26:24.000000000 -0500
+++ ../../lib/php/jaxl/core/jaxl.parser.php	2012-04-12 16:14:46.677746486 -0500
@@ -67,7 +67,8 @@
             ),
         
             'success'       =>  array(
-                'xmlns'     =>  '//success/@xmlns'
+                'xmlns'     =>  '//success/@xmlns',
+                'success'   =>  '//success/text()'
             ),
             
             'failure'       =>  array(
diff -ru -x .gitignore ../abhinavsingh-JAXL-5829c3b/env/jaxl.ini ../../lib/php/jaxl/env/jaxl.ini
--- ../abhinavsingh-JAXL-5829c3b/env/jaxl.ini	2011-05-01 07:26:24.000000000 -0500
+++ ../../lib/php/jaxl/env/jaxl.ini	2012-04-12 17:53:47.075183594 -0500
@@ -50,9 +50,6 @@
      * Values specified here are overwritten by those passed with JAXL constructor
     */  
     
-    // PHP reporting level
-    error_reporting(E_ERROR | E_WARNING | E_PARSE);
-    
     // Connecting user credentials
     define('JAXL_USER_NAME', 'user');
     define('JAXL_USER_PASS', 'pass');
diff -ru -x .gitignore ../abhinavsingh-JAXL-5829c3b/xmpp/xmpp.auth.php ../../lib/php/jaxl/xmpp/xmpp.auth.php
--- ../abhinavsingh-JAXL-5829c3b/xmpp/xmpp.auth.php	2011-05-01 07:26:24.000000000 -0500
+++ ../../lib/php/jaxl/xmpp/xmpp.auth.php	2012-04-12 14:57:55.098352802 -0500
@@ -114,32 +114,8 @@
                 $xml .= base64_encode(JAXLUtil::implodeData($response));
             }
             else if($authType == 'SCRAM-SHA-1') {
-                $decoded = JAXLUtil::explodeData($decoded);
-                        
-                // SaltedPassword  := Hi(Normalize(password), salt, i)
-                $saltedPasswd = JAXLUtil::pbkdf2($jaxl->pass, $decoded['s'], $decoded['i']);
-                        
-                // ClientKey       := HMAC(SaltedPassword, "Client Key")
-                $clientKey = JAXLUtil::hashMD5($saltedPassword, "Client Key");
-                        
-                // StoredKey       := H(ClientKey)
-                $storedKey = sha1("Client Key");
-                        
-                // assemble client-final-message-without-proof
-                $clientFinalMessage = "c=bwis,r=".$decoded['r'];
-                        
-                // AuthMessage     := client-first-message-bare + "," + server-first-message + "," + client-final-message-without-proof
-                // ClientSignature := HMAC(StoredKey, AuthMessage)
-                
-                // ClientProof     := ClientKey XOR ClientSignature
-                // ServerKey       := HMAC(SaltedPassword, "Server Key")
-                // ServerSignature := HMAC(ServerKey, AuthMessage)
-                        
-                foreach(array('c', 'r', 'p') as $key)
-                    if(isset($decoded[$key]))
-                        $response[$key] = $decoded[$key];
-                        
-                $xml .= base64_encode(JAXLUtil::implodeData($response));
+                $response = $jaxl->scram->getResponse($jaxl->user, $jaxl->pass, $decoded);
+                $xml .= base64_encode($response);
             }
             else if($authType == 'CRAM-MD5') {
                 $xml .= base64_encode($jaxl->user.' '.hash_hmac('md5', $jaxl->pass, $arr['challenge']));
diff -ru -x .gitignore ../abhinavsingh-JAXL-5829c3b/xmpp/xmpp.get.php ../../lib/php/jaxl/xmpp/xmpp.get.php
--- ../abhinavsingh-JAXL-5829c3b/xmpp/xmpp.get.php	2011-05-01 07:26:24.000000000 -0500
+++ ../../lib/php/jaxl/xmpp/xmpp.get.php	2012-04-12 14:59:40.572560233 -0500
@@ -145,6 +145,11 @@
         }
         
         public static function success($arr, $jaxl) {
+            if($jaxl->authType == "SCRAM-SHA-1" && !$jaxl->scram->processOutcome(base64_decode($arr['success']))) {
+                $jaxl->log("[[XMPPGet]] SCRAM-SHA-1 server verification failed");
+                self::failure($arr, $jaxl);
+                return;
+            }
             if($arr['xmlns'] == "urn:ietf:params:xml:ns:xmpp-sasl") {
                 if($jaxl->mode == "cgi") JAXL0206::restartStream($jaxl);
                 else XMPPSend::startStream($jaxl);
diff -ru -x .gitignore ../abhinavsingh-JAXL-5829c3b/xmpp/xmpp.send.php ../../lib/php/jaxl/xmpp/xmpp.send.php
--- ../abhinavsingh-JAXL-5829c3b/xmpp/xmpp.send.php	2011-05-01 07:26:24.000000000 -0500
+++ ../../lib/php/jaxl/xmpp/xmpp.send.php	2012-04-12 14:55:00.061060489 -0500
@@ -86,7 +86,10 @@
                 case 'CRAM-MD5':
                     break;
                 case 'SCRAM-SHA-1':
-                    $xml .= base64_encode("n,,n=".$jaxl->user.",r=".base64_encode(JAXLUtil::generateNonce()));
+                    // PEAR Auth_SASL2 Library
+                    require_once("Auth/SASL2/SCRAM.php");
+                    $jaxl->scram = new Auth_SASL2_SCRAM("SHA-1");
+                    $xml .= base64_encode($jaxl->scram->getResponse($jaxl->user, $jaxl->pass));
                     break;
                 default:
                     break;
